From 37bfa214ba33d4ca13816cc4c95594f4770e3223 Mon Sep 17 00:00:00 2001
From: WANG Xuerui <git@xen0n.name>
Date: Mon, 17 May 2021 17:58:16 +0800
Subject: [PATCH 18/19] Revert "1.Add vadd/vmadd/vmul/vsub back to insn 2.Delet
 greymark insn 3.Delet some insn except jump insn"

This reverts commit ac08c97529c6a540e8345de631f5348072a2018e.

There is no evidence these instructions actually disappeared, so restore
them for experiments.
---
 opcodes/loongarch-opc.c | 309 ++++++++++++++++++++++++++++------------
 1 file changed, 214 insertions(+), 95 deletions(-)

diff --git a/opcodes/loongarch-opc.c b/opcodes/loongarch-opc.c
index 83f8103d9a6..893a79a524a 100644
--- a/opcodes/loongarch-opc.c
+++ b/opcodes/loongarch-opc.c
@@ -290,8 +290,8 @@ static struct loongarch_opcode loongarch_fix_opcodes[] = {
 {0x00006c00, 0xfffffc00, "cpucfg", "r0:5,r5:5", 0, 0, 0, 0},
 {0x00007000, 0xffffff1f, "x86mttop", "u5:3", 0, 0, 0, 0},
 {0x00007400, 0xffffffe0, "x86mftop", "r0:5", 0, 0, 0, 0},
-{0x00007800, 0xfffffc00, "setx86loope", "r0:5,r5:5", 0, 0, 0, 0},
-{0x00007c00, 0xfffffc00, "setx86loopne", "r0:5,r5:5", 0, 0, 0, 0},
+{0x00007800, 0xfffffc00, "x86loope", "r0:5,r5:5", 0, 0, 0, 0},
+{0x00007c00, 0xfffffc00, "x86loopne", "r0:5,r5:5", 0, 0, 0, 0},
 {0x00008000, 0xfffffc1f, "x86inc.b", "r5:5", 0, 0, 0, 0},
 {0x00008001, 0xfffffc1f, "x86inc.h", "r5:5", 0, 0, 0, 0},
 {0x00008002, 0xfffffc1f, "x86inc.w", "r5:5", 0, 0, 0, 0},
@@ -524,49 +524,71 @@ static struct loongarch_opcode loongarch_fix_opcodes[] = {
 
 static struct loongarch_opcode loongarch_float_opcodes[] = {
 /* match,    mask,       name, format, macro, include, exclude, pinfo */
+{0x01000000, 0xffff8000, "fadd.h", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
 {0x01008000, 0xffff8000, "fadd.s", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
 {0x01010000, 0xffff8000, "fadd.d", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
+{0x01020000, 0xffff8000, "fsub.h", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
 {0x01028000, 0xffff8000, "fsub.s", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
 {0x01030000, 0xffff8000, "fsub.d", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
+{0x01040000, 0xffff8000, "fmul.h", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
 {0x01048000, 0xffff8000, "fmul.s", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
 {0x01050000, 0xffff8000, "fmul.d", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
+{0x01060000, 0xffff8000, "fdiv.h", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
 {0x01068000, 0xffff8000, "fdiv.s", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
 {0x01070000, 0xffff8000, "fdiv.d", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
+{0x01080000, 0xffff8000, "fmax.h", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
 {0x01088000, 0xffff8000, "fmax.s", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
 {0x01090000, 0xffff8000, "fmax.d", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
+{0x010a0000, 0xffff8000, "fmin.h", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
 {0x010a8000, 0xffff8000, "fmin.s", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
 {0x010b0000, 0xffff8000, "fmin.d", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
+{0x010c0000, 0xffff8000, "fmaxa.h", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
 {0x010c8000, 0xffff8000, "fmaxa.s", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
 {0x010d0000, 0xffff8000, "fmaxa.d", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
+{0x010e0000, 0xffff8000, "fmina.h", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
 {0x010e8000, 0xffff8000, "fmina.s", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
 {0x010f0000, 0xffff8000, "fmina.d", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
+{0x01100000, 0xffff8000, "fscaleb.h", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
 {0x01108000, 0xffff8000, "fscaleb.s", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
 {0x01110000, 0xffff8000, "fscaleb.d", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
+{0x01120000, 0xffff8000, "fcopysign.h", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
 {0x01128000, 0xffff8000, "fcopysign.s", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
 {0x01130000, 0xffff8000, "fcopysign.d", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
+{0x01140000, 0xfffffc00, "fabs.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01140400, 0xfffffc00, "fabs.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01140800, 0xfffffc00, "fabs.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01141000, 0xfffffc00, "fneg.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01141400, 0xfffffc00, "fneg.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01141800, 0xfffffc00, "fneg.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01142000, 0xfffffc00, "flogb.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01142400, 0xfffffc00, "flogb.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01142800, 0xfffffc00, "flogb.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01143000, 0xfffffc00, "fclass.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01143400, 0xfffffc00, "fclass.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01143800, 0xfffffc00, "fclass.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01144000, 0xfffffc00, "fsqrt.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01144400, 0xfffffc00, "fsqrt.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01144800, 0xfffffc00, "fsqrt.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01145000, 0xfffffc00, "frecip.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01145400, 0xfffffc00, "frecip.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01145800, 0xfffffc00, "frecip.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01146000, 0xfffffc00, "frsqrt.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01146400, 0xfffffc00, "frsqrt.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01146800, 0xfffffc00, "frsqrt.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01147000, 0xfffffc00, "frecipe.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01147400, 0xfffffc00, "frecipe.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01147800, 0xfffffc00, "frecipe.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01148000, 0xfffffc00, "frsqrte.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01148400, 0xfffffc00, "frsqrte.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01148800, 0xfffffc00, "frsqrte.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01149000, 0xfffffc00, "fmov.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01149400, 0xfffffc00, "fmov.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01149800, 0xfffffc00, "fmov.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x0114a000, 0xfffffc00, "movgr2fr.h", "f0:5,r5:5", 0, 0, 0, 0},
 {0x0114a400, 0xfffffc00, "movgr2fr.w", "f0:5,r5:5", 0, 0, 0, 0},
 {0x0114a800, 0xfffffc00, "movgr2fr.d", "f0:5,r5:5", 0, 0, 0, 0},
 {0x0114ac00, 0xfffffc00, "movgr2frh.w", "f0:5,r5:5", 0, 0, 0, 0},
+{0x0114b000, 0xfffffc00, "movfr2gr.h", "r0:5,f5:5", 0, 0, 0, 0},
 {0x0114b400, 0xfffffc00, "movfr2gr.s", "r0:5,f5:5", 0, 0, 0, 0},
 {0x0114b800, 0xfffffc00, "movfr2gr.d", "r0:5,f5:5", 0, 0, 0, 0},
 {0x0114bc00, 0xfffffc00, "movfrh2gr.s", "r0:5,f5:5", 0, 0, 0, 0},
@@ -581,88 +603,244 @@ static struct loongarch_opcode loongarch_float_opcodes[] = {
 {0x01150000, 0xffff8000, "fcvt.d.ld", "f0:5,f5:5,f10:5", 0, 0, 0, 0},
 {0x01180400, 0xfffffc00, "fcvtrm.h.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01180800, 0xfffffc00, "fcvtrm.h.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01180c00, 0xfffffc00, "fcvtrm.h.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01181000, 0xfffffc00, "fcvtrm.s.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01181800, 0xfffffc00, "fcvtrm.s.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01181c00, 0xfffffc00, "fcvtrm.s.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01182000, 0xfffffc00, "fcvtrm.d.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01182400, 0xfffffc00, "fcvtrm.d.s", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01182c00, 0xfffffc00, "fcvtrm.d.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01183000, 0xfffffc00, "fcvtrm.q.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01183400, 0xfffffc00, "fcvtrm.q.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01183800, 0xfffffc00, "fcvtrm.q.d", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01184400, 0xfffffc00, "fcvtrp.h.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01184800, 0xfffffc00, "fcvtrp.h.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01184c00, 0xfffffc00, "fcvtrp.h.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01185000, 0xfffffc00, "fcvtrp.s.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01185800, 0xfffffc00, "fcvtrp.s.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01185c00, 0xfffffc00, "fcvtrp.s.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01186000, 0xfffffc00, "fcvtrp.d.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01186400, 0xfffffc00, "fcvtrp.d.s", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01186c00, 0xfffffc00, "fcvtrp.d.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01187000, 0xfffffc00, "fcvtrp.q.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01187400, 0xfffffc00, "fcvtrp.q.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01187800, 0xfffffc00, "fcvtrp.q.d", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01188400, 0xfffffc00, "fcvtrz.h.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01188800, 0xfffffc00, "fcvtrz.h.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01188c00, 0xfffffc00, "fcvtrz.h.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01189000, 0xfffffc00, "fcvtrz.s.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01189800, 0xfffffc00, "fcvtrz.s.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01189c00, 0xfffffc00, "fcvtrz.s.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x0118a000, 0xfffffc00, "fcvtrz.d.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x0118a400, 0xfffffc00, "fcvtrz.d.s", "f0:5,f5:5", 0, 0, 0, 0},
+{0x0118ac00, 0xfffffc00, "fcvtrz.d.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x0118b000, 0xfffffc00, "fcvtrz.q.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x0118b400, 0xfffffc00, "fcvtrz.q.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x0118b800, 0xfffffc00, "fcvtrz.q.d", "f0:5,f5:5", 0, 0, 0, 0},
 {0x0118c400, 0xfffffc00, "fcvtrne.h.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x0118c800, 0xfffffc00, "fcvtrne.h.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x0118cc00, 0xfffffc00, "fcvtrne.h.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x0118d000, 0xfffffc00, "fcvtrne.s.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x0118d800, 0xfffffc00, "fcvtrne.s.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x0118dc00, 0xfffffc00, "fcvtrne.s.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x0118e000, 0xfffffc00, "fcvtrne.d.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x0118e400, 0xfffffc00, "fcvtrne.d.s", "f0:5,f5:5", 0, 0, 0, 0},
+{0x0118ec00, 0xfffffc00, "fcvtrne.d.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x0118f000, 0xfffffc00, "fcvtrne.q.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x0118f400, 0xfffffc00, "fcvtrne.q.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x0118f800, 0xfffffc00, "fcvtrne.q.d", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01190400, 0xfffffc00, "fcvt.h.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01190800, 0xfffffc00, "fcvt.h.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01190c00, 0xfffffc00, "fcvt.h.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01191000, 0xfffffc00, "fcvt.s.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01191800, 0xfffffc00, "fcvt.s.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01191c00, 0xfffffc00, "fcvt.s.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01192000, 0xfffffc00, "fcvt.d.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01192400, 0xfffffc00, "fcvt.d.s", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01192c00, 0xfffffc00, "fcvt.d.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x01193000, 0xfffffc00, "fcvt.q.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01193400, 0xfffffc00, "fcvt.q.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x01193800, 0xfffffc00, "fcvt.q.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011a0000, 0xfffffc00, "ftintrm.w.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011a0400, 0xfffffc00, "ftintrm.w.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011a0800, 0xfffffc00, "ftintrm.w.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011a0c00, 0xfffffc00, "ftintrm.w.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011a1000, 0xfffffc00, "ftintrm.wu.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011a1400, 0xfffffc00, "ftintrm.wu.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011a1800, 0xfffffc00, "ftintrm.wu.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011a1c00, 0xfffffc00, "ftintrm.wu.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011a2000, 0xfffffc00, "ftintrm.l.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011a2400, 0xfffffc00, "ftintrm.l.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011a2800, 0xfffffc00, "ftintrm.l.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011a2c00, 0xfffffc00, "ftintrm.l.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011a3000, 0xfffffc00, "ftintrm.lu.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011a3400, 0xfffffc00, "ftintrm.lu.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011a3800, 0xfffffc00, "ftintrm.lu.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011a3c00, 0xfffffc00, "ftintrm.lu.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011a4000, 0xfffffc00, "ftintrp.w.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011a4400, 0xfffffc00, "ftintrp.w.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011a4800, 0xfffffc00, "ftintrp.w.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011a4c00, 0xfffffc00, "ftintrp.w.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011a5000, 0xfffffc00, "ftintrp.wu.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011a5400, 0xfffffc00, "ftintrp.wu.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011a5800, 0xfffffc00, "ftintrp.wu.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011a5c00, 0xfffffc00, "ftintrp.wu.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011a6000, 0xfffffc00, "ftintrp.l.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011a6400, 0xfffffc00, "ftintrp.l.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011a6800, 0xfffffc00, "ftintrp.l.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011a6c00, 0xfffffc00, "ftintrp.l.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011a7000, 0xfffffc00, "ftintrp.lu.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011a7400, 0xfffffc00, "ftintrp.lu.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011a7800, 0xfffffc00, "ftintrp.lu.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011a7c00, 0xfffffc00, "ftintrp.lu.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011a8000, 0xfffffc00, "ftintrz.w.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011a8400, 0xfffffc00, "ftintrz.w.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011a8800, 0xfffffc00, "ftintrz.w.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011a8c00, 0xfffffc00, "ftintrz.w.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011a9000, 0xfffffc00, "ftintrz.wu.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011a9400, 0xfffffc00, "ftintrz.wu.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011a9800, 0xfffffc00, "ftintrz.wu.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011a9c00, 0xfffffc00, "ftintrz.wu.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011aa000, 0xfffffc00, "ftintrz.l.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011aa400, 0xfffffc00, "ftintrz.l.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011aa800, 0xfffffc00, "ftintrz.l.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011aac00, 0xfffffc00, "ftintrz.l.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011ab000, 0xfffffc00, "ftintrz.lu.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011ab400, 0xfffffc00, "ftintrz.lu.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011ab800, 0xfffffc00, "ftintrz.lu.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011abc00, 0xfffffc00, "ftintrz.lu.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011ac000, 0xfffffc00, "ftintrne.w.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011ac400, 0xfffffc00, "ftintrne.w.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011ac800, 0xfffffc00, "ftintrne.w.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011acc00, 0xfffffc00, "ftintrne.w.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011ad000, 0xfffffc00, "ftintrne.wu.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011ad400, 0xfffffc00, "ftintrne.wu.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011ad800, 0xfffffc00, "ftintrne.wu.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011adc00, 0xfffffc00, "ftintrne.wu.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011ae000, 0xfffffc00, "ftintrne.l.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011ae400, 0xfffffc00, "ftintrne.l.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011ae800, 0xfffffc00, "ftintrne.l.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011aec00, 0xfffffc00, "ftintrne.l.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011af000, 0xfffffc00, "ftintrne.lu.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011af400, 0xfffffc00, "ftintrne.lu.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011af800, 0xfffffc00, "ftintrne.lu.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011afc00, 0xfffffc00, "ftintrne.lu.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011b0000, 0xfffffc00, "ftint.w.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011b0400, 0xfffffc00, "ftint.w.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011b0800, 0xfffffc00, "ftint.w.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011b0c00, 0xfffffc00, "ftint.w.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011b1000, 0xfffffc00, "ftint.wu.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011b1400, 0xfffffc00, "ftint.wu.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011b1800, 0xfffffc00, "ftint.wu.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011b1c00, 0xfffffc00, "ftint.wu.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011b2000, 0xfffffc00, "ftint.l.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011b2400, 0xfffffc00, "ftint.l.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011b2800, 0xfffffc00, "ftint.l.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011b2c00, 0xfffffc00, "ftint.l.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011b3000, 0xfffffc00, "ftint.lu.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011b3400, 0xfffffc00, "ftint.lu.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011b3800, 0xfffffc00, "ftint.lu.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011b3c00, 0xfffffc00, "ftint.lu.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c0000, 0xfffffc00, "ffintrm.h.w", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c0400, 0xfffffc00, "ffintrm.h.wu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c0800, 0xfffffc00, "ffintrm.h.l", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c0c00, 0xfffffc00, "ffintrm.h.lu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c1000, 0xfffffc00, "ffintrm.s.w", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c1400, 0xfffffc00, "ffintrm.s.wu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c1800, 0xfffffc00, "ffintrm.s.l", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c1c00, 0xfffffc00, "ffintrm.s.lu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c2000, 0xfffffc00, "ffintrm.d.w", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c2400, 0xfffffc00, "ffintrm.d.wu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c2800, 0xfffffc00, "ffintrm.d.l", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c2c00, 0xfffffc00, "ffintrm.d.lu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c3000, 0xfffffc00, "ffintrm.q.w", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c3400, 0xfffffc00, "ffintrm.q.wu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c3800, 0xfffffc00, "ffintrm.q.l", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c3c00, 0xfffffc00, "ffintrm.q.lu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c4000, 0xfffffc00, "ffintrp.h.w", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c4400, 0xfffffc00, "ffintrp.h.wu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c4800, 0xfffffc00, "ffintrp.h.l", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c4c00, 0xfffffc00, "ffintrp.h.lu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c5000, 0xfffffc00, "ffintrp.s.w", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c5400, 0xfffffc00, "ffintrp.s.wu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c5800, 0xfffffc00, "ffintrp.s.l", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c5c00, 0xfffffc00, "ffintrp.s.lu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c6000, 0xfffffc00, "ffintrp.d.w", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c6400, 0xfffffc00, "ffintrp.d.wu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c6800, 0xfffffc00, "ffintrp.d.l", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c6c00, 0xfffffc00, "ffintrp.d.lu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c7000, 0xfffffc00, "ffintrp.q.w", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c7400, 0xfffffc00, "ffintrp.q.wu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c7800, 0xfffffc00, "ffintrp.q.l", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c7c00, 0xfffffc00, "ffintrp.q.lu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c8000, 0xfffffc00, "ffintrz.h.w", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c8400, 0xfffffc00, "ffintrz.h.wu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c8800, 0xfffffc00, "ffintrz.h.l", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c8c00, 0xfffffc00, "ffintrz.h.lu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c9000, 0xfffffc00, "ffintrz.s.w", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c9400, 0xfffffc00, "ffintrz.s.wu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c9800, 0xfffffc00, "ffintrz.s.l", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011c9c00, 0xfffffc00, "ffintrz.s.lu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011ca000, 0xfffffc00, "ffintrz.d.w", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011ca400, 0xfffffc00, "ffintrz.d.wu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011ca800, 0xfffffc00, "ffintrz.d.l", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011cac00, 0xfffffc00, "ffintrz.d.lu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011cb000, 0xfffffc00, "ffintrz.q.w", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011cb400, 0xfffffc00, "ffintrz.q.wu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011cb800, 0xfffffc00, "ffintrz.q.l", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011cbc00, 0xfffffc00, "ffintrz.q.lu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011cc000, 0xfffffc00, "ffintrne.h.w", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011cc400, 0xfffffc00, "ffintrne.h.wu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011cc800, 0xfffffc00, "ffintrne.h.l", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011ccc00, 0xfffffc00, "ffintrne.h.lu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011cd000, 0xfffffc00, "ffintrne.s.w", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011cd400, 0xfffffc00, "ffintrne.s.wu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011cd800, 0xfffffc00, "ffintrne.s.l", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011cdc00, 0xfffffc00, "ffintrne.s.lu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011ce000, 0xfffffc00, "ffintrne.d.w", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011ce400, 0xfffffc00, "ffintrne.d.wu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011ce800, 0xfffffc00, "ffintrne.d.l", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011cec00, 0xfffffc00, "ffintrne.d.lu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011cf000, 0xfffffc00, "ffintrne.q.w", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011cf400, 0xfffffc00, "ffintrne.q.wu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011cf800, 0xfffffc00, "ffintrne.q.l", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011cfc00, 0xfffffc00, "ffintrne.q.lu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011d0000, 0xfffffc00, "ffint.h.w", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011d0400, 0xfffffc00, "ffint.h.wu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011d0800, 0xfffffc00, "ffint.h.l", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011d0c00, 0xfffffc00, "ffint.h.lu", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011d1000, 0xfffffc00, "ffint.s.w", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011d1400, 0xfffffc00, "ffint.s.wu", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011d1800, 0xfffffc00, "ffint.s.l", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011d1c00, 0xfffffc00, "ffint.s.lu", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011d2000, 0xfffffc00, "ffint.d.w", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011d2400, 0xfffffc00, "ffint.d.wu", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011d2800, 0xfffffc00, "ffint.d.l", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011d2c00, 0xfffffc00, "ffint.d.lu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011d3000, 0xfffffc00, "ffint.q.w", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011d3400, 0xfffffc00, "ffint.q.wu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011d3800, 0xfffffc00, "ffint.q.l", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011d3c00, 0xfffffc00, "ffint.q.lu", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011e0000, 0xfffffc00, "frintrm.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011e0400, 0xfffffc00, "frintrm.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011e0800, 0xfffffc00, "frintrm.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011e0c00, 0xfffffc00, "frintrm.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011e1000, 0xfffffc00, "frintrp.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011e1400, 0xfffffc00, "frintrp.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011e1800, 0xfffffc00, "frintrp.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011e1c00, 0xfffffc00, "frintrp.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011e2000, 0xfffffc00, "frintrz.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011e2400, 0xfffffc00, "frintrz.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011e2800, 0xfffffc00, "frintrz.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011e2c00, 0xfffffc00, "frintrz.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011e3000, 0xfffffc00, "frintrne.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011e3400, 0xfffffc00, "frintrne.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011e3800, 0xfffffc00, "frintrne.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011e3c00, 0xfffffc00, "frintrne.q", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011e4000, 0xfffffc00, "frint.h", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011e4400, 0xfffffc00, "frint.s", "f0:5,f5:5", 0, 0, 0, 0},
 {0x011e4800, 0xfffffc00, "frint.d", "f0:5,f5:5", 0, 0, 0, 0},
+{0x011e4c00, 0xfffffc00, "frint.q", "f0:5,f5:5", 0, 0, 0, 0},
 {0} /* Terminate the list.  */
 };
 
@@ -719,6 +897,7 @@ static struct loongarch_opcode loongarch_privilege_opcodes[] = {
 {0x06483400, 0xffffffff, "tlbfill", "", 0, 0, 0, 0},
 {0x06483401, 0xffffffff, "gtlbfill", "", 0, 0, 0, 0},
 {0x06483800, 0xffffffff, "ertn", "", 0, 0, 0, 0},
+{0x06483c00, 0xffffffff, "deret", "", 0, 0, 0, 0},
 {0x06488000, 0xffff8000, "idle", "u0:15", 0, 0, 0, 0},
 {0x06498000, 0xffff8000, "invtlb", "u0:5,r5:5,r10:5", 0, 0, 0, 0},
 {0} /* Terminate the list.  */
@@ -726,30 +905,50 @@ static struct loongarch_opcode loongarch_privilege_opcodes[] = {
 
 static struct loongarch_opcode loongarch_4opt_opcodes[] = {
 /* match,    mask,       name, format, macro, include, exclude, pinfo */
+{0x08000000, 0xfff00000, "fmadd.h", "f0:5,f5:5,f10:5,f15:5", 0, 0, 0, 0},
 {0x08100000, 0xfff00000, "fmadd.s", "f0:5,f5:5,f10:5,f15:5", 0, 0, 0, 0},
 {0x08200000, 0xfff00000, "fmadd.d", "f0:5,f5:5,f10:5,f15:5", 0, 0, 0, 0},
+{0x08400000, 0xfff00000, "fmsub.h", "f0:5,f5:5,f10:5,f15:5", 0, 0, 0, 0},
 {0x08500000, 0xfff00000, "fmsub.s", "f0:5,f5:5,f10:5,f15:5", 0, 0, 0, 0},
 {0x08600000, 0xfff00000, "fmsub.d", "f0:5,f5:5,f10:5,f15:5", 0, 0, 0, 0},
+{0x08800000, 0xfff00000, "fnmadd.h", "f0:5,f5:5,f10:5,f15:5", 0, 0, 0, 0},
 {0x08900000, 0xfff00000, "fnmadd.s", "f0:5,f5:5,f10:5,f15:5", 0, 0, 0, 0},
 {0x08a00000, 0xfff00000, "fnmadd.d", "f0:5,f5:5,f10:5,f15:5", 0, 0, 0, 0},
+{0x08c00000, 0xfff00000, "fnmsub.h", "f0:5,f5:5,f10:5,f15:5", 0, 0, 0, 0},
 {0x08d00000, 0xfff00000, "fnmsub.s", "f0:5,f5:5,f10:5,f15:5", 0, 0, 0, 0},
 {0x08e00000, 0xfff00000, "fnmsub.d", "f0:5,f5:5,f10:5,f15:5", 0, 0, 0, 0},
+{0x09000000, 0xfff00000, "vfmadd.h", "v0:5,v5:5,v10:5,v15:5", 0, 0, 0, 0},
 {0x09100000, 0xfff00000, "vfmadd.s", "v0:5,v5:5,v10:5,v15:5", 0, 0, 0, 0},
 {0x09200000, 0xfff00000, "vfmadd.d", "v0:5,v5:5,v10:5,v15:5", 0, 0, 0, 0},
+{0x09300000, 0xfff00000, "vfmadd.q", "v0:5,v5:5,v10:5,v15:5", 0, 0, 0, 0},
+{0x09400000, 0xfff00000, "vfmsub.h", "v0:5,v5:5,v10:5,v15:5", 0, 0, 0, 0},
 {0x09500000, 0xfff00000, "vfmsub.s", "v0:5,v5:5,v10:5,v15:5", 0, 0, 0, 0},
 {0x09600000, 0xfff00000, "vfmsub.d", "v0:5,v5:5,v10:5,v15:5", 0, 0, 0, 0},
+{0x09700000, 0xfff00000, "vfmsub.q", "v0:5,v5:5,v10:5,v15:5", 0, 0, 0, 0},
+{0x09800000, 0xfff00000, "vfnmadd.h", "v0:5,v5:5,v10:5,v15:5", 0, 0, 0, 0},
 {0x09900000, 0xfff00000, "vfnmadd.s", "v0:5,v5:5,v10:5,v15:5", 0, 0, 0, 0},
 {0x09a00000, 0xfff00000, "vfnmadd.d", "v0:5,v5:5,v10:5,v15:5", 0, 0, 0, 0},
+{0x09b00000, 0xfff00000, "vfnmadd.q", "v0:5,v5:5,v10:5,v15:5", 0, 0, 0, 0},
+{0x09c00000, 0xfff00000, "vfnmsub.h", "v0:5,v5:5,v10:5,v15:5", 0, 0, 0, 0},
 {0x09d00000, 0xfff00000, "vfnmsub.s", "v0:5,v5:5,v10:5,v15:5", 0, 0, 0, 0},
 {0x09e00000, 0xfff00000, "vfnmsub.d", "v0:5,v5:5,v10:5,v15:5", 0, 0, 0, 0},
+{0x09f00000, 0xfff00000, "vfnmsub.q", "v0:5,v5:5,v10:5,v15:5", 0, 0, 0, 0},
+{0x0a000000, 0xfff00000, "xvfmadd.h", "x0:5,x5:5,x10:5,x15:5", 0, 0, 0, 0},
 {0x0a100000, 0xfff00000, "xvfmadd.s", "x0:5,x5:5,x10:5,x15:5", 0, 0, 0, 0},
 {0x0a200000, 0xfff00000, "xvfmadd.d", "x0:5,x5:5,x10:5,x15:5", 0, 0, 0, 0},
+{0x0a300000, 0xfff00000, "xvfmadd.q", "x0:5,x5:5,x10:5,x15:5", 0, 0, 0, 0},
+{0x0a400000, 0xfff00000, "xvfmsub.h", "x0:5,x5:5,x10:5,x15:5", 0, 0, 0, 0},
 {0x0a500000, 0xfff00000, "xvfmsub.s", "x0:5,x5:5,x10:5,x15:5", 0, 0, 0, 0},
 {0x0a600000, 0xfff00000, "xvfmsub.d", "x0:5,x5:5,x10:5,x15:5", 0, 0, 0, 0},
+{0x0a700000, 0xfff00000, "xvfmsub.q", "x0:5,x5:5,x10:5,x15:5", 0, 0, 0, 0},
+{0x0a800000, 0xfff00000, "xvfnmadd.h", "x0:5,x5:5,x10:5,x15:5", 0, 0, 0, 0},
 {0x0a900000, 0xfff00000, "xvfnmadd.s", "x0:5,x5:5,x10:5,x15:5", 0, 0, 0, 0},
 {0x0aa00000, 0xfff00000, "xvfnmadd.d", "x0:5,x5:5,x10:5,x15:5", 0, 0, 0, 0},
+{0x0ab00000, 0xfff00000, "xvfnmadd.q", "x0:5,x5:5,x10:5,x15:5", 0, 0, 0, 0},
+{0x0ac00000, 0xfff00000, "xvfnmsub.h", "x0:5,x5:5,x10:5,x15:5", 0, 0, 0, 0},
 {0x0ad00000, 0xfff00000, "xvfnmsub.s", "x0:5,x5:5,x10:5,x15:5", 0, 0, 0, 0},
 {0x0ae00000, 0xfff00000, "xvfnmsub.d", "x0:5,x5:5,x10:5,x15:5", 0, 0, 0, 0},
+{0x0af00000, 0xfff00000, "xvfnmsub.q", "x0:5,x5:5,x10:5,x15:5", 0, 0, 0, 0},
 {0x0c000000, 0xffff8018, "fcmp.caf.h", "c0:3,f5:5,f10:5", 0, 0, 0, 0},
 {0x0c008000, 0xffff8018, "fcmp.saf.h", "c0:3,f5:5,f10:5", 0, 0, 0, 0},
 {0x0c010000, 0xffff8018, "fcmp.clt.h", "c0:3,f5:5,f10:5", 0, 0, 0, 0},
@@ -1008,14 +1207,24 @@ static struct loongarch_opcode loongarch_4opt_opcodes[] = {
 {0x0d200000, 0xfff00000, "xvbitsel.v", "x0:5,x5:5,x10:5,x15:5", 0, 0, 0, 0},
 {0x0d500000, 0xfff00000, "vshuf.b", "v0:5,v5:5,v10:5,v15:5", 0, 0, 0, 0},
 {0x0d600000, 0xfff00000, "xvshuf.b", "x0:5,x5:5,x10:5,x15:5", 0, 0, 0, 0},
+{0x0d900000, 0xfff00000, "vextr.v", "v0:5,v5:5,v10:5,u15:5", 0, 0, 0, 0},
+{0x0da00000, 0xfff00000, "xvextr.v", "x0:5,x5:5,x10:5,u15:5", 0, 0, 0, 0},
+{0x0e800000, 0xfff00000, "vfmaddsub.h", "v0:5,v5:5,v10:5,v15:5", 0, 0, 0, 0},
 {0x0e900000, 0xfff00000, "vfmaddsub.s", "v0:5,v5:5,v10:5,v15:5", 0, 0, 0, 0},
 {0x0ea00000, 0xfff00000, "vfmaddsub.d", "v0:5,v5:5,v10:5,v15:5", 0, 0, 0, 0},
+{0x0eb00000, 0xfff00000, "vfmaddsub.q", "v0:5,v5:5,v10:5,v15:5", 0, 0, 0, 0},
+{0x0ec00000, 0xfff00000, "vfmsubadd.h", "v0:5,v5:5,v10:5,v15:5", 0, 0, 0, 0},
 {0x0ed00000, 0xfff00000, "vfmsubadd.s", "v0:5,v5:5,v10:5,v15:5", 0, 0, 0, 0},
 {0x0ee00000, 0xfff00000, "vfmsubadd.d", "v0:5,v5:5,v10:5,v15:5", 0, 0, 0, 0},
+{0x0ef00000, 0xfff00000, "vfmsubadd.q", "v0:5,v5:5,v10:5,v15:5", 0, 0, 0, 0},
+{0x0f000000, 0xfff00000, "xvfmaddsub.h", "x0:5,x5:5,x10:5,x15:5", 0, 0, 0, 0},
 {0x0f100000, 0xfff00000, "xvfmaddsub.s", "x0:5,x5:5,x10:5,x15:5", 0, 0, 0, 0},
 {0x0f200000, 0xfff00000, "xvfmaddsub.d", "x0:5,x5:5,x10:5,x15:5", 0, 0, 0, 0},
+{0x0f300000, 0xfff00000, "xvfmaddsub.q", "x0:5,x5:5,x10:5,x15:5", 0, 0, 0, 0},
+{0x0f400000, 0xfff00000, "xvfmsubadd.h", "x0:5,x5:5,x10:5,x15:5", 0, 0, 0, 0},
 {0x0f500000, 0xfff00000, "xvfmsubadd.s", "x0:5,x5:5,x10:5,x15:5", 0, 0, 0, 0},
 {0x0f600000, 0xfff00000, "xvfmsubadd.d", "x0:5,x5:5,x10:5,x15:5", 0, 0, 0, 0},
+{0x0f700000, 0xfff00000, "xvfmsubadd.q", "x0:5,x5:5,x10:5,x15:5", 0, 0, 0, 0},
 {0} /* Terminate the list.  */
 };
 
@@ -1761,7 +1970,10 @@ static struct loongarch_opcode loongarch_128vec_opcodes[] = {
 {0x73408000, 0xffff8000, "vsrlni.h.w", "v0:5,v5:5,u10:5", 0, 0, 0, 0},
 {0x73410000, 0xffff0000, "vsrlni.w.d", "v0:5,v5:5,u10:6", 0, 0, 0, 0},
 {0x73420000, 0xfffe0000, "vsrlni.d.q", "v0:5,v5:5,u10:7", 0, 0, 0, 0},
+{0x73444000, 0xffffc000, "vsrlrni.b.h", "v0:5,v5:5,u10:4", 0, 0, 0, 0},
+{0x73448000, 0xffff8000, "vsrlrni.h.w", "v0:5,v5:5,u10:5", 0, 0, 0, 0},
 {0x73450000, 0xffff0000, "vsrlrni.w.d", "v0:5,v5:5,u10:6", 0, 0, 0, 0},
+{0x73460000, 0xfffe0000, "vsrlrni.d.q", "v0:5,v5:5,u10:7", 0, 0, 0, 0},
 {0x73484000, 0xffffc000, "vssrlni.b.h", "v0:5,v5:5,u10:4", 0, 0, 0, 0},
 {0x73488000, 0xffff8000, "vssrlni.h.w", "v0:5,v5:5,u10:5", 0, 0, 0, 0},
 {0x73490000, 0xffff0000, "vssrlni.w.d", "v0:5,v5:5,u10:6", 0, 0, 0, 0},
@@ -1816,99 +2028,6 @@ static struct loongarch_opcode loongarch_128vec_opcodes[] = {
 {0x73d80000, 0xfffc0000, "vxori.b", "v0:5,v5:5,u10:8", 0, 0, 0, 0},
 {0x73dc0000, 0xfffc0000, "vnori.b", "v0:5,v5:5,u10:8", 0, 0, 0, 0},
 {0, 0, "vrepli.b", "v,s0:10", "vldi %1,(%2)&0x3ff", 0, 0, 0},
-
-{0x701e0000, 0xffff8000, "vaddwev.h.b", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x701e8000, 0xffff8000, "vaddwev.w.h", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x701f0000, 0xffff8000, "vaddwev.d.w", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x701f8000, 0xffff8000, "vaddwev.q.d", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x702e0000, 0xffff8000, "vaddwev.h.bu", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x702e8000, 0xffff8000, "vaddwev.w.hu", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x702f0000, 0xffff8000, "vaddwev.d.wu", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x702f8000, 0xffff8000, "vaddwev.q.du", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x703e0000, 0xffff8000, "vaddwev.h.bu.b", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x703e8000, 0xffff8000, "vaddwev.w.hu.h", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x703f0000, 0xffff8000, "vaddwev.d.wu.w", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x703f8000, 0xffff8000, "vaddwev.q.du.d", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70220000, 0xffff8000, "vaddwod.h.b", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70228000, 0xffff8000, "vaddwod.w.h", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70230000, 0xffff8000, "vaddwod.d.w", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70238000, 0xffff8000, "vaddwod.q.d", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70320000, 0xffff8000, "vaddwod.h.bu", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70328000, 0xffff8000, "vaddwod.w.hu", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70330000, 0xffff8000, "vaddwod.d.wu", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70338000, 0xffff8000, "vaddwod.q.du", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70400000, 0xffff8000, "vaddwod.h.bu.b", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70408000, 0xffff8000, "vaddwod.w.hu.h", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70410000, 0xffff8000, "vaddwod.d.wu.w", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70418000, 0xffff8000, "vaddwod.q.du.d", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70ac0000, 0xffff8000, "vmaddwev.h.b", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70ac8000, 0xffff8000, "vmaddwev.w.h", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70ad0000, 0xffff8000, "vmaddwev.d.w", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70ad8000, 0xffff8000, "vmaddwev.q.d", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70b40000, 0xffff8000, "vmaddwev.h.bu", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70b48000, 0xffff8000, "vmaddwev.w.hu", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70b50000, 0xffff8000, "vmaddwev.d.wu", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70b58000, 0xffff8000, "vmaddwev.q.du", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70bc0000, 0xffff8000, "vmaddwev.h.bu.b", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70bc8000, 0xffff8000, "vmaddwev.w.hu.h", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70bd0000, 0xffff8000, "vmaddwev.d.wu.w", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70bd8000, 0xffff8000, "vmaddwev.q.du.d", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70ae0000, 0xffff8000, "vmaddwod.h.b", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70ae8000, 0xffff8000, "vmaddwod.w.h", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70af0000, 0xffff8000, "vmaddwod.d.w", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70af8000, 0xffff8000, "vmaddwod.q.d", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70b60000, 0xffff8000, "vmaddwod.h.bu", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70b68000, 0xffff8000, "vmaddwod.w.hu", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70b70000, 0xffff8000, "vmaddwod.d.wu", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70b78000, 0xffff8000, "vmaddwod.q.du", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70be0000, 0xffff8000, "vmaddwod.h.bu.b", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70be8000, 0xffff8000, "vmaddwod.w.hu.h", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70bf0000, 0xffff8000, "vmaddwod.d.wu.w", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70bf8000, 0xffff8000, "vmaddwod.q.du.d", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70900000, 0xffff8000, "vmulwev.h.b", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70908000, 0xffff8000, "vmulwev.w.h", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70910000, 0xffff8000, "vmulwev.d.w", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70918000, 0xffff8000, "vmulwev.q.d", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70980000, 0xffff8000, "vmulwev.h.bu", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70988000, 0xffff8000, "vmulwev.w.hu", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70990000, 0xffff8000, "vmulwev.d.wu", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70998000, 0xffff8000, "vmulwev.q.du", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70a00000, 0xffff8000, "vmulwev.h.bu.b", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70a08000, 0xffff8000, "vmulwev.w.hu.h", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70a10000, 0xffff8000, "vmulwev.d.wu.w", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70a18000, 0xffff8000, "vmulwev.q.du.d", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70920000, 0xffff8000, "vmulwod.h.b", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70928000, 0xffff8000, "vmulwod.w.h", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70930000, 0xffff8000, "vmulwod.d.w", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70938000, 0xffff8000, "vmulwod.q.d", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x709a0000, 0xffff8000, "vmulwod.h.bu", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x709a8000, 0xffff8000, "vmulwod.w.hu", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x709b0000, 0xffff8000, "vmulwod.d.wu", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x709b8000, 0xffff8000, "vmulwod.q.du", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70a20000, 0xffff8000, "vmulwod.h.bu.b", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70a28000, 0xffff8000, "vmulwod.w.hu.h", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70a30000, 0xffff8000, "vmulwod.d.wu.w", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70a38000, 0xffff8000, "vmulwod.q.du.d", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70200000, 0xffff8000, "vsubwev.h.b", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70208000, 0xffff8000, "vsubwev.w.h", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70210000, 0xffff8000, "vsubwev.d.w", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70218000, 0xffff8000, "vsubwev.q.d", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70300000, 0xffff8000, "vsubwev.h.bu", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70308000, 0xffff8000, "vsubwev.w.hu", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70310000, 0xffff8000, "vsubwev.d.wu", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70318000, 0xffff8000, "vsubwev.q.du", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70240000, 0xffff8000, "vsubwod.h.b", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70248000, 0xffff8000, "vsubwod.w.h", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70250000, 0xffff8000, "vsubwod.d.w", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70258000, 0xffff8000, "vsubwod.q.d", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70340000, 0xffff8000, "vsubwod.h.bu", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70348000, 0xffff8000, "vsubwod.w.hu", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70350000, 0xffff8000, "vsubwod.d.wu", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-{0x70358000, 0xffff8000, "vsubwod.q.du", "v0:5,v5:5,v10:5", 0, 0, 0, 0},
-
-
-
-
 {0, 0, "vrepli.d", "v,s0:10", "vldi %1,((%2)&0x3ff)|0xc00", 0, 0, 0},
 {0, 0, "vrepli.h", "v,s0:10", "vldi %1,((%2)&0x3ff)|0x400", 0, 0, 0},
 {0, 0, "vrepli.w", "v,s0:10", "vldi %1,((%2)&0x3ff)|0x800", 0, 0, 0},
-- 
2.30.1

